core/lib/zksync_core/src/api_server/execution_sandbox/apply.rs

```diff
@@ -36,7 +37,250 @@ use super::{
+        // We don't want to emit too many logs.
+        if resolve_time > Duration::from_millis(10) {
+            tracing::debug!("Resolved block numbers (took {resolve_time:?})");
+        }
+
+        if block_args.resolves_to_latest_sealed_miniblock() {
+            shared_args
+                .caches
+                .schedule_values_update(resolved_block_info.state_l2_block_number);
+        }
+
+        let (next_l2_block_info, l2_block_info_to_reset) = Self::load_l2_block_info(
+            &mut connection,
+            block_args.is_pending_miniblock(),
+            &resolved_block_info,
+        )
+        .await?;
+
+        let storage = PostgresStorage::new_async(
+            Handle::current(),
+            connection,
+            resolved_block_info.state_l2_block_number,
+            false,
+        )
+        .await
+        .context("cannot create `PostgresStorage`")?
+        .with_caches(shared_args.caches.clone());
+
+        let storage_view = StorageView::new(storage);
+        let (system_env, l1_batch_env) = Self::prepare_env(
+            shared_args,
+            execution_args,
+            &resolved_block_info,
+            next_l2_block_info,
+        );
+
+        Ok(Self {
+            system_env,
+            l1_batch_env,
+            storage_view,
+            execution_args,
+            l2_block_info_to_reset,
+        })
+    }
+
+    async fn load_l2_block_info(
+        connection: &mut StorageProcessor<'_>,
+        is_pending_block: bool,
+        resolved_block_info: &ResolvedBlockInfo,
+    ) -> anyhow::Result<(L2BlockEnv, Option<StoredL2BlockInfo>)> {
+        let mut l2_block_info_to_reset = None;
+        let current_l2_block_info = StoredL2BlockInfo::new(
+            connection,
+            resolved_block_info.state_l2_block_number,
+            Some(resolved_block_info.state_l2_block_hash),
+        )
+        .await
+        .context("failed reading L2 block info")?;
+
+        let next_l2_block_info = if is_pending_block {
+            L2BlockEnv {
+                number: current_l2_block_info.l2_block_number + 1,
+                timestamp: resolved_block_info.l1_batch_timestamp,
+                prev_block_hash: current_l2_block_info.l2_block_hash,
+                // For simplicity we assume each miniblock create one virtual block.
+                // This may be wrong only during transition period.
+                max_virtual_blocks_to_create: 1,
+            }
+        } else if current_l2_block_info.l2_block_number == 0 {
+            // Special case:
+            // - For environments, where genesis block was created before virtual block upgrade it doesn't matter what we put here.
+            // - Otherwise, we need to put actual values here. We cannot create next L2 block with block_number=0 and `max_virtual_blocks_to_create=0`
+            //   because of SystemContext requirements. But, due to intrinsics of SystemContext, block.number still will be resolved to 0.
```

Comment:
given that virtual blocks migration is over, can this be simplified? @slowli good that you extracted this method, but the logic is still a little hard to follow

Reply:
As you've said, all work on `Sandbox` is essentially just moving code around. I'd prefer to do any semantics changes separately, so it'd be easier to review and revert if necessary.